File: C Sharp/Blink/Blink/Async/AsyncReceiveDispatcher.cs
using Net.Qiujuer.Blink.Core;
using System;
using System.Net.Sockets;

namespace Net.Qiujuer.Blink.Async
{
    /// <summary>
    /// Provides for performing receive dispatch from a queue of BinkConn {@link BlinkConn}.
    /// </summary>
    public class AsyncReceiveDispatcher : AsyncAbsDispatcher
    {
        /// <summary>
        /// Socket receive buffer size
        /// </summary>
        private readonly int mBufferSize;
        /// <summary>
        /// Parser the receive data type
        /// </summary>
        private BlinkParser mParser;
        /// <summary>
        /// Receive Data
        /// </summary>
        private Receiver mReceiver;
        /// <summary>
        /// Posting responses.
        /// </summary>
        private ReceiveDelivery mReceiveDelivery;
        private BlinkDelivery mBlinkDelivery;

        private ReceivePacket mReceivePacket;
        private short mSurplusInfoLen;
        private long mSurplusLen;


        public AsyncReceiveDispatcher(Receiver receiver, BlinkParser parser, ReceiveDelivery receiveDelivery, BlinkDelivery blinkDelivery, float progressPrecision)
            : base(progressPrecision)
        {
            mReceiver = receiver;
            mBufferSize = receiver.GetBufferSize();
            mParser = parser;
            mReceiveDelivery = receiveDelivery;
            mBlinkDelivery = blinkDelivery;

            // Set Buffer
            SetBuffer(new byte[mBufferSize], 0, mBufferSize);

            // Start
            ReceiveAsync(0);
        }

        private void ReceiveAsync(long size)
        {
            int count;
            if (size > mBufferSize)
                count = mBufferSize;
            else if (size <= 0)
                count = HeadSize;
            else
                count = Convert.ToInt32(size);

            ReceiveAsync(0, count);
        }

        private void ReceiveAsync(int offset, int count)
        {
            if (mDisposed)
                return;

            // Set post buffer
            SetBuffer(offset, count);

            // Post a receive to the connection
            mStatus = mReceiver.ReceiveAsync(this);
            if (!mStatus)
            {
                // On sync call
                OnCompleted(this);
            }
        }

        private void ReceiveHead(byte[] buffer)
        {
            mSurplusLen = 0;
            mSurplusInfoLen = 0;
            mProgress = 0;

            byte type = buffer[0];
            long len = BitConverter.ToInt64(buffer, 1);
            short info = BitConverter.ToInt16(buffer, HeadSize - 2);

            if (len > 0)
            {
                // Set Length
                mSurplusLen = len;
                mSurplusInfoLen = info;

                // Parse receive packet
                ReceivePacket packet = mParser.ParseReceive(type, len);

                if (packet != null && packet.StartPacket())
                {
                    mReceivePacket = packet;

                    // Notifly
                    ReceiveDelivery delivery = mReceiveDelivery;
                    if (delivery != null)
                        delivery.PostReceiveStart(packet);

                }
                else
                {
                    // Set Null
                    mReceivePacket = null;
                }
            }

            ReceiveAsync(mSurplusInfoLen > 0 ? mSurplusInfoLen : mSurplusLen);
        }

        private void ReceiveInfo(byte[] buffer, int offset, int count)
        {
            // Set len
            mSurplusInfoLen -= Convert.ToInt16(count);

            if (mSurplusInfoLen > 0)
            {
                // Receive info
                ReceiveAsync(offset + count, mSurplusInfoLen);
            }
            else
            {
                // Set Info
                ReceivePacket packet = mReceivePacket;
                if (packet != null)
                {
                    packet.WriteInfo(buffer, 0, offset + count);
                }

                // Receive entity
                ReceiveAsync(mSurplusLen);
            }
        }

        private void ReceiveEntity(byte[] buffer, int offset, int count)
        {
            // Set len
            mSurplusLen -= count;

            ReceivePacket packet = mReceivePacket;

            if (packet != null)
            {
                packet.Write(buffer, offset, count);


                // Notity progress
                float len = packet.GetLength();
                float progress = (len - mSurplusLen) / len;

                // Post Callback
                ReceiveDelivery delivery = mReceiveDelivery;

                if (IsNotifyProgress(progress))
                {
                    // Notify                    
                    if (delivery != null)
                        delivery.PostReceiveProgress(packet, mProgress);
                }

                if (mSurplusLen <= 0)
                {
                    // End
                    packet.EndPacket();

                    // Notify
                    if (delivery != null)
                        delivery.PostReceiveEnd(packet, mStatus);

                    // Set Null
                    mReceivePacket = null;
                }
            }

            // Receive next entity
            ReceiveAsync(mSurplusLen);
        }

        protected override void OnCompleted(SocketAsyncEventArgs e)
        {
            // Call base
            base.OnCompleted(e);

            // Check if the remote host closed the connection
            if (e.LastOperation == SocketAsyncOperation.Receive
                && e.BytesTransferred > 0
                && e.SocketError == SocketError.Success)
            {
                // Receive Entity
                if (mSurplusInfoLen > 0)
                    ReceiveInfo(e.Buffer, e.Offset, e.BytesTransferred);
                else if (mSurplusLen > 0)
                    ReceiveEntity(e.Buffer, e.Offset, e.BytesTransferred);
                else
                {
                    if (e.BytesTransferred < e.Count)
                        // Full the head
                        ReceiveAsync(e.Offset + e.BytesTransferred, e.Count - e.BytesTransferred);
                    else
                        // Receive Head
                        ReceiveHead(e.Buffer);
                }
            }
            else
            {
                Dispose();
            }
        }

        public new void Dispose()
        {
            if (!IsDisposed())
            {
                mParser = null;

                ReceivePacket packet = mReceivePacket;
                mReceivePacket = null;

                Receiver receiver = mReceiver;
                mReceiver = null;

                ReceiveDelivery receiveDelivery = mReceiveDelivery;
                mReceiveDelivery = null;

                if (packet != null && receiveDelivery != null)
                {
                    if (mSurplusLen > 0)
                    {
                        packet.EndPacket();
                        receiveDelivery.PostReceiveEnd(packet, false);
                    }
                }

                BlinkDelivery blinkDelay = mBlinkDelivery;
                mBlinkDelivery = null;
                if (blinkDelay != null)
                    blinkDelay.PostBlinkDisconnect();

                if (receiver != null)
                    receiver.Dispose();

                SetBuffer(null, 0, 0);

                base.Dispose();
            }
        }
    }
}

File: Java/Blink/Blink/src/net/qiujuer/blink/async/AsyncReceiveDispatcher.java
/*
 * Copyright (C) 2014 Qiujuer <qiujuer@live.cn>
 * WebSite http://www.qiujuer.net
 * Created 04/16/2015
 * Changed 04/26/2015
 * Version 1.0.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.qiujuer.blink.async;

import net.qiujuer.blink.BlinkClient;
import net.qiujuer.blink.core.Connector;
import net.qiujuer.blink.core.PacketFilter;
import net.qiujuer.blink.core.PacketParser;
import net.qiujuer.blink.core.ReceiveDispatcher;
import net.qiujuer.blink.core.ReceivePacket;
import net.qiujuer.blink.core.Receiver;
import net.qiujuer.blink.core.delivery.ReceiveDelivery;

/**
 * Provides for performing receive dispatch from a queue of BinkConn {@link BlinkClient}.
 */
public class AsyncReceiveDispatcher extends AsyncDispatcher implements ReceiveDispatcher {
    // Parser the receive data type
    private PacketParser mParser;
    // Receive Data
    private Receiver mReceiver;
    // Posting responses.
    private ReceiveDelivery mReceiveDelivery;
    // Blink Connector
    private Connector mConnector;


    public AsyncReceiveDispatcher(Receiver receiver, PacketParser parser, Connector connector, ReceiveDelivery receiveDelivery) {
        // Set Buffer
        super(receiver.getReceiveBufferSize(), connector.getProgressPrecision());

        mReceiver = receiver;
        mParser = parser;
        mReceiveDelivery = receiveDelivery;
        mConnector = connector;

        // Start
        mParser.setEventArgs(this);
        receiveAsync();
    }

    /**
     * Start async receive
     */
    private void receiveAsync() {
        if (mDisposed.get())
            return;
        // As soon as the client is connected, post a receive to the connection
        mReceiver.receiveAsync(this);
    }

    /**
     * Receive packet by parser
     * Notify send progress
     */
    private void receivePacket() {
        float progress = mParser.parse();
        // Notify
        ReceiveDelivery delivery = mReceiveDelivery;
        if (delivery != null && isNotifyProgress(progress)) {
            if (progress == PacketFilter.STATUS_START) {
                delivery.postReceiveStart(mConnector, mParser.getPacket());
            } else if (progress == PacketFilter.STATUS_END) {
                delivery.postReceiveCompleted(mConnector, mParser.getPacket());
            } else {
                delivery.postReceiveProgress(mConnector, mParser.getPacket(), progress);
            }
        }
    }

    /**
     * On async receive completed
     *
     * @param e IoEventArgs {@link IoEventArgs}
     */
    @Override
    protected void onCompleted(IoEventArgs e) {
        // Check if the remote host closed the connection
        int transferred = e.getBytesTransferred();
        if (transferred > 0) {
            if (transferred < e.getCount()) {
                // Full the head
                e.setBuffer(e.getOffset() + transferred, e.getCount() - transferred);
                receiveAsync();
            } else {
                receivePacket();
                receiveAsync();
            }
        } else {
            dispose();
        }
    }

    @Override
    public void dispose() {
        if (mDisposed.compareAndSet(false, true)) {
            ReceivePacket packet = mParser.getPacket();
            if (packet != null)
                packet.endPacket();
            mParser = null;
            mReceiver = null;
            mReceiveDelivery = null;
        }
    }
}
